### ISSUE TITLE
Add `AppStartUp` UserTag to track timing of Flutter app start

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/flutter#84884]
Soon, DDS is going to support caching CPU profiles for a given user tag and provide an API for Flutter DevTools to request a cached profile. @bkonyi is working on this. We will use this functionality to support profiling app startup for Flutter apps. In order to use this functionality, we need a [UserTag](https://api.flutter.dev/flutter/dart-developer/UserTag-class.html) `AppStartUp` to be set at the beginning of app start, and unset at the end. This will mark the start and end times for app startup in the CPU profiler. 

I'm not sure where in the flutter ecosystem these UserTag changes will need to take place. If either of them need to happen in Native code in the engine, @bkonyi can make a small change to the C API to support this. @chinmaygarde @iskakaushik

Flutter Tools will then need to pass `AppStartUp`, or whatever the name of the user tag ends up being, as a flag value to DDS, so that DDS knows to cache profiles for that tag. @jonahwilliams

**Comment by jonahwilliams** [comment]
If we're passing through the app startup arguments, this will mean that app start timing will only work when flutter launches the app and not through any sort of attach workflows. I don't have any concerns with this, just bringing it up as a possible issue.

Otherwise its fairly easy to pass args to the engine, but then the different engine embeddings would need to either recognize them _or_ we could pass through the existing --dart-flags option, provided they are added to the allowlist in https://github.com/flutter/engine/blob/master/shell/common/switches.cc#L43-L54

**Comment by kenzieschmoll** [comment]
Candidate locations to set/unset the AppStartUp tags as recommended by @xster 

set:

> https://github.com/flutter/engine/blob/master/runtime/dart_vm.cc#L448 (though it might be a bit too early) or https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/binding.dart#L53 (though it might be a bit too late). 

I would rather error on the side of too early than too late. @bkonyi we would need to expose the dart developer api to the C code somehow to go with the dart_vm.cc location.

unset: https://github.com/flutter/flutter/blob/4ba8835d77ba14cdfe80456979626347f977c5fb/packages/flutter/lib/src/widgets/binding.dart#L863

**Comment by bkonyi** [comment]
@kenzieschmoll [UserTag support landed a few weeks ago in the SDK](https://github.com/dart-lang/sdk/blob/master/runtime/include/dart_tools_api.h#L501).

**Comment by kenzieschmoll** [comment]
> UserTag support landed a few weeks ago in the SDK.

Oh, great!

@iskakaushik, would you be able to make the proposed change to the engine here https://github.com/flutter/engine/blob/master/runtime/dart_vm.cc#L448? The Dart equivalent would be `UserTag('AppStartUp').makeCurrent()`.

I can make the change to the flutter framework to reset the UserTag to the default tag.

**Comment by iskakaushik** [comment]
@kenzieschmoll, totally missed this sorry, looks like you're already working on a PR: https://github.com/flutter/engine/pull/27553, that change LGTM.

**Comment by zanderso** [comment]
Looking at https://github.com/flutter/flutter/pull/86516, will DevTools get confused if a Flutter developer calls `UserTag.makeCurrent`  on some other tag in e.g. `main()` before the first frame goes up? If there is a public API that is legal for a developer to call that will cause problems, then this approach seems like it might be fragile. Do we have documentation somewhere asking Flutter devs not to set the current user tag?

**Comment by kenzieschmoll** [comment]
We could add a check in the framework that only sets the default user tag if the current tag is "AppStartUp". It would also be nice if there was a place to keep the "AppStartUp" string in a const that is accessible by both the engine and the framework so we don't have fragile string matching

**Comment by zanderso** [comment]
From our meeting just now:
1. We should add a check to the framework side of this that only overwrites the user tag to be the default again if it is "AppStartUp".
2. The "AppStartUp" tag string can be stored in `dart:ui`.
3. The string "AppStartUp" can be passed to dds from the Flutter tool as a hardcoded string with a link to the dart:ui code.

Also the documentation should note that this won't include any samples from before Dart `main()` starts running.

**Comment by iskakaushik** [comment]
> The "AppStartUp" tag string can be stored in dart:ui.

@zanderso on further thought, given that on the engine side this constant will only be used (once) in C++ code, I don't see the value of putting this in `dart:ui`, we could probably just put this as a constant in the framework if we think it will be used multiple times.

**Comment by zanderso** [comment]
Yeah, making it a public API of dart:ui so that it would accessible to both the engine C++ code and the framework code would be a little bit special-purpose. We'll have to make sure that there is mutual linking in comments to the locations of the hard-coded strings, then.

**Comment by jmagman** [comment]
flutter/flutter#89600 caused many Android tests to time out or OOM.  I was able to reproduce locally with `complex_layout_android__scroll_smoothness`.

https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8836611845424144241/+/u/run_hot_mode_dev_cycle_linux__benchmark/test_stdout
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8836611845870259377/+/u/run_complex_layout_android__scroll_smoothness/test_stdout
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8836611848881226001/+/u/run_channels_integration_test_win/test_stdout
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8836611846766817729/+/u/run_drive_perf_debug_warning/test_stdout

Reverted at https://github.com/flutter/flutter/pull/89704.

**Comment by bkonyi** [comment]
Looks like we're trying to process CPU samples on the UI thread which is causing some extreme jank. I'll investigate.

![image](https://user-images.githubusercontent.com/24210656/132702210-b630f8ed-7d1c-484a-9dff-0cce08a60bb8.png)

**Comment by kenzieschmoll** [comment]
@iskakaushik would it be possible to set the `--profile_period` flag on the VM at startup to a high sampling rate? I'm seeing the issue where we only sometimes have samples for the "AppStartUp" user tag, and I believe this is due to the fact that the default sampling rate is 1000Hz (once per millisecond). Can we change this to a high sampling rate of 50hz (once per 50 microseconds) - this is what DevTools uses for the high sampling rate setting. Then we can reset the sampling rate to the default value when we unset the AppStartUp user tag here: https://github.com/flutter/flutter/blob/33f4107cd1029c3847f119e23719f6147dbc39db/packages/flutter/lib/src/widgets/binding.dart#L866l (that is, if we have access to the vm service from here to call `_vmService.setFlag`).

In general, can we make the default sampling rate higher? as soon as DevTools is opened, we change the sampling rate to 250 (once per 250 micros), anyway. Can we change the default value to match this? @bkonyi @zanderso
